### PR TITLE
#28 profile activity category models

### PR DESCRIPTION
**Pull request by ddianaoo**
Add models and serializers for Profile, Category and Activity

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -110,7 +110,12 @@ $ bower install
 ### How to run local
 
 ### How to run Docker
-
+- install Docker
+- make sure you put your cradentials in .env
+- use command in terminal: docker-compose up -d --build
+- running your containers
+- follow the link
+- in some cases of mistakes you need to put line 30 in settings.py in comment and uncomment line 31, then delete containers and images and repeat 3, 4, 5 steps
```
Comment (mehalyna) [inline README.md:118]: add steps counting to the items, pls

File: profiles/models.py [hunk 2]
```diff
@@ -1,3 +1,55 @@
...
 
-# Create your models here.
+from validation.validate_password import validate_password
+from validation.validate_phone_number import validate_phone_number_len, validate_phone_number_is_digit
+from validation.validate_edrpou import validate_edrpou
+from validation.validate_foundation_year import validate_foundation_year
+from validation.validate_image import validate_image_size, validate_image_format
+
+#REGIONS = ["east", "west", "north", "south"]
+REGIONS = (('E', "east"), ('W', "west"), ('N', "north"), ('S', "south"))
```
Comment (mehalyna) [inline profiles/models.py:11]: isn't it more expedient to use a dictionary here?

File: profiles/models.py [hunk 3]
```diff
@@ -1,3 +1,55 @@
...
+            validate_phone_number_len,
+            RegexValidator(r"^\d+$", message="Phone number must contain only numbers."),
+        ]
+    )
+    comp_EDRPOU = models.IntegerField(unique=True, validators=[validate_edrpou])
+    comp_year_of_foundation = models.SmallIntegerField(validators=[validate_foundation_year])
+    comp_service_info = models.TextField()
+    comp_product_info = models.TextField()
+    comp_address = models.TextField()
+    comp_registered = models.BinaryField(max_length=1)
```
Comment (mehalyna) [inline profiles/models.py:35]: what is the purpose of this attribute?

File: validation/validate_foundation_year.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+from django.core.exceptions import ValidationError
+
+
+def validate_foundation_year(foundation_year_value: int):
+    if len(str(foundation_year_value)) != 4:
```
Comment (mehalyna) [inline validation/validate_foundation_year.py:5]: and what if foundation_year_value > current_year ?

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: profiles/models.py (hunk 3 above)
Comment (mehalyna) [reply profiles/models.py:35]: yep

File: README.md (hunk 1 above)
Comment (BelousSofiya) [reply README.md]: Thank you for feedback. Done

File: profiles/models.py (hunk 3 above)
Comment (mehalyna) [reply profiles/models.py:35]: Sorry for the possible misunderstanding, but it looks like we can use only one switch on/off in comp_startup. Although, with the help of two attributes at the same time, we can cover the requirement when the subject can be both a startup and a company at the same time. Leave it for now.

File: profiles/models.py (hunk 2 above)
Comment (BelousSofiya) [reply profiles/models.py:11]: As we found, we have to use only tuple of tuples for field attribute 'choices'. It's on line 21 models.py. Are we wrong?

File: profiles/models.py (hunk 3 above)
Comment (BelousSofiya) [reply profiles/models.py:35]: Ok, we will think about this problem. Thank you for feedback.

File: validation/validate_foundation_year.py (hunk 4 above)
Comment (ddianaoo) [reply validation/validate_foundation_year.py:8]: corrected, thank you!

Comment (eduard999m) [reply validation/validate_foundation_year.py:8]: Done

File: profiles/models.py (hunk 2 above)
Comment (mehalyna) [reply profiles/models.py:12]: indeed :(
as soon as we need choices here. let it be up to the next improvement.

File: profiles/models.py [hunk 5]
```diff
@@ -1,3 +1,55 @@
...
+from django.core.validators import RegexValidator
 
-# Create your models here.
+from validation.validate_password import validate_password
+from validation.validate_phone_number import validate_phone_number_len, validate_phone_number_is_digit
+from validation.validate_edrpou import validate_edrpou
+from validation.validate_foundation_year import validate_foundation_year
+from validation.validate_image import validate_image_size, validate_image_format
+
+#REGIONS = ["east", "west", "north", "south"]
```
Comment (mehalyna) [inline profiles/models.py:11]: Drop this line, pls.

No one commented out line of code in repo!

File: profiles/models.py (hunk 2 above)
Comment (BelousSofiya) [reply profiles/models.py:12]: Thank you we'll try to find better solution for the next time.

File: validation/validate_image.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+def validate_image_format(image: Image):
+    valid_formats = ["png", "jpeg", "jpg"]
+    img = Image.open(image)
+    format_ = img.format.lower()
+    if format_ not in valid_formats:
+        raise ValidationError("Unsupported image format. Only PNG and JPEG are allowed.")
+
+
+def validate_image_size(image: Image):
```
[the remark this answers is not in the document]
Comment (BelousSofiya) [reply validation/validate_image.py:13]: We'll leave this for the next time/ Thank you for interesting idea))

**Review by Lvyshnevska — APPROVED** [review]

File: validation/validate_foundation_year.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+def validate_foundation_year_range(foundation_year_value: int) -> None:
+    if not (foundation_year_value in range(1800, (datetime.datetime.now().year+1))):
+        raise ValidationError("Foundation year must be exactly in range 1800-current year.")
+
+
+def validate_foundation_year_is_valid(foundation_year_value: int) -> None:
+    if foundation_year_value > datetime.datetime.now().year:
+        raise ValidationError("Foundation year cannot be more than current year.")
+
```
Comment (Shushunya) [inline validation/validate_foundation_year.py:11]: Unneccessary method.

Comment (BelousSofiya) [reply validation/validate_foundation_year.py:11]: Delete.

File: validation/validate_foundation_year.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+def validate_foundation_year(foundation_year_value: int):
+    current_year = datetime.datetime.now().year
+    if len(str(foundation_year_value)) != 4 or foundation_year_value > current_year:
+>>>>>>> 56283f56932a50fbaa281beb0044fa670cc029ec
+        raise ValidationError("Foundation year must be exactly 4 digits long.")
+
+
+def validate_foundation_year_is_valid(foundation_year_value: int) -> None:
```
[the remark this answers is not in the document]
Comment (BelousSofiya) [reply validation/validate_foundation_year.py]: Delete.

File: validation/validate_foundation_year.py [hunk 9]
```diff
@@ -1,10 +1,19 @@
...
     if len(str(foundation_year_value)) != 4:
+=======
+from django.core.exceptions import ValidationError
+
+
+def validate_foundation_year(foundation_year_value: int):
+    current_year = datetime.datetime.now().year
+    if len(str(foundation_year_value)) != 4 or foundation_year_value > current_year:
+>>>>>>> 56283f56932a50fbaa281beb0044fa670cc029ec
         raise ValidationError("Foundation year must be exactly 4 digits long.")
```
[the remark this answers is not in the document]
Comment (ddianaoo) [reply validation/validate_foundation_year.py]: we have already changed this method. Thank you for your note

File: validation/validate_phone_number.py [hunk 10]
```diff
@@ -0,0 +1,11 @@
+from django.core.exceptions import ValidationError
+
+
+def validate_phone_number_len(phone_number_value: str):
```
[the remark this answers is not in the document]
Comment (ddianaoo) [reply validation/validate_phone_number.py:4]: We decided to leave validation methods for phone number and split password validation into two methods

**Review by Shushunya — APPROVED** [review]

**Pull request merged**